File: functions/getCheckouts.js
// functions/getCheckouts.js

const { CookieJar } = await import('tough-cookie');
const fetchModule = await import('node-fetch');
const fetchCookieModule = await import('fetch-cookie');
const fetch = fetchModule.default;
const fetchCookie = fetchCookieModule.default;

const jar = new CookieJar();
const fetchWithCookies = fetchCookie(fetch, jar);

const LOGIN_URL     = "https://aclibrary.bibliocommons.com/user/login?destination=%2Faccount%2Fcontact_preferences";
const CHECKOUTS_URL = "https://gateway.bibliocommons.com/v2/libraries/aclibrary/checkouts";

exports.handler = async (event) => {
  const { CookieJar } = await import('tough-cookie');
  const fetchModule = await import('node-fetch');
  const fetchCookieModule = await import('fetch-cookie');
  const fetch = fetchModule.default;
  const fetchCookie = fetchCookieModule.default;

  const jar = new CookieJar();
  const fetchWithCookies = fetchCookie(fetch, jar);

  const { name, user_pin, accountId } = event.queryStringParameters;

  const LOGIN_URL     = "https://aclibrary.bibliocommons.com/user/login?destination=%2Faccount%2Fcontact_preferences";
  const CHECKOUTS_URL = "https://gateway.bibliocommons.com/v2/libraries/aclibrary/checkouts";

  await fetchWithCookies(LOGIN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8' },
    body: new URLSearchParams({ name, user_pin })
  });

  const url = `${CHECKOUTS_URL}?accountId=${accountId}&size=100&page=1&status=OUT&sort=status&locale=en-US`;
  const resp = await fetchWithCookies(url, { headers: { Accept: 'application/json' } });
  const data = await resp.json();

  return {
    statusCode: 200,
    headers: { 'Access-Control-Allow-Origin': '*' },
    body: JSON.stringify(data),
  };
};

File: functions/get_checkouts.py
# functions/get_checkouts.py

import json
import urllib.parse
import requests

LOGIN_URL     = "https://aclibrary.bibliocommons.com/user/login?destination=%2Faccount%2Fcontact_preferences"
CHECKOUTS_URL = "https://gateway.bibliocommons.com/v2/libraries/aclibrary/checkouts"

def handler(event, context):
    # Extract query parameters
    params = event.get("queryStringParameters") or {}
    name = params.get("name")
    user_pin = params.get("user_pin")
    account_id = params.get("accountId")

    if not all([name, user_pin, account_id]):
        return {
            "statusCode": 400,
            "body": json.dumps({ "error": "Missing name, user_pin, or accountId" }),
        }

    try:
        # Use a session to maintain cookies
        session = requests.Session()

        # Perform login
        payload = { "name": name, "user_pin": user_pin }
        login_resp = session.post(
            LOGIN_URL,
            headers={ "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8" },
            data=urllib.parse.urlencode(payload)
        )
        login_resp.raise_for_status()

        # Fetch checkouts
        params = {
            "accountId": account_id,
            "size":      100,
            "status":    "OUT",
            "page":      1,
            "sort":      "status",
            "locale":    "en-US"
        }
        r = session.get(
            CHECKOUTS_URL,
            headers={ "Accept": "application/json" },
            params=params
        )
        r.raise_for_status()

        return {
            "statusCode": 200,
            "headers": { "Access-Control-Allow-Origin": "*" },
            "body": json.dumps(r.json())
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({ "error": str(e) })
        }
